agent/agent.go

```diff
@@ -650,46 +644,39 @@ func (a *Agent) listenAndServeGRPC() error {
 }
 
 func (a *Agent) listenAndServeDNS() error {
-	notif := make(chan net.Addr, len(a.config.DNSAddrs))
-	errCh := make(chan error, len(a.config.DNSAddrs))
+	chNotify := make(chan struct{}, len(a.config.DNSAddrs))
+
 	for _, addr := range a.config.DNSAddrs {
-		// create server
 		s, err := NewDNSServer(a)
 		if err != nil {
 			return err
 		}
-		a.dnsServers = append(a.dnsServers, s)
 
-		// start server
-		a.wgServers.Add(1)
-		go func(addr net.Addr) {
-			defer a.wgServers.Done()
-			err := s.ListenAndServe(addr.Network(), addr.String(), func() { notif <- addr })
-			if err != nil && !strings.Contains(err.Error(), "accept") {
-				errCh <- err
-			}
-		}(addr)
+		done := func() {
+			chNotify <- struct{}{}
+		}
+		a.apiServers.Start(newAPIServerDNS(s, addr, done))
+		a.configReloaders = append(a.configReloaders, reloadConfigDNSServer(s))
 	}
 
-	// wait for servers to be up
+	// Unlike the HTTP server, the dns.Server listens and serves in the same call.
+	// We could fix this by calling ActivateAndServe instead of ListenAndServe,
+	// however that would require duplicating the network switch/case in
+	// ListenAndServe. Do we actually support more than one network type?
+	// For now we are handling this difference by waiting up to a second for
+	// the listener to listen.
+	// TODO: fix this by using dns.Server.ActivateAndServe
```

Comment:
This was the old behaviour as well. I've kept it for now, but I think we can change it in the future. It'll be much easier to test with the DNSServer is better isolated from Agent.